src/compiler/binder.ts

```diff
@@ -1751,4 +1791,437 @@ namespace ts {
+
+                // If a FunctionExpression contains an asterisk token, or its subtree has marked the container
+                // as needing to capture the lexical this, then this node is ES6 syntax.
+                if ((<FunctionLikeDeclaration>node).asteriskToken
+                    || subtreeFlags & TransformFlags.ContainsCapturedLexicalThis
+                    || subtreeFlags & TransformFlags.ContainsDefaultValueAssignments) {
+                    transformFlags |= TransformFlags.AssertES6;
+                }
+
+                // An async function expression is TypeScript syntax.
+                if (node.flags & NodeFlags.Async) {
+                    transformFlags |= TransformFlags.AssertTypeScript;
+                }
+
+                break;
+
+            case SyntaxKind.FunctionDeclaration:
+                // A FunctionDeclaration excludes markers that should not escape the scope of a FunctionDeclaration.
+                excludeFlags = TransformFlags.FunctionExcludes;
+
+                // A FunctionDeclaration without a body is an overload and is TypeScript syntax.
+                if (!(<FunctionDeclaration>node).body) {
+                    transformFlags = TransformFlags.AssertTypeScript;
+                    break;
+                }
+
+                // If a FunctionDeclaration has an asterisk token, is exported, or its
+                // subtree has marked the container as needing to capture the lexical `this`,
+                // then this node is ES6 syntax.
+                if ((<FunctionDeclaration>node).asteriskToken
+                    || node.flags & NodeFlags.Export
+                    || subtreeFlags & TransformFlags.ContainsCapturedLexicalThis
+                    || subtreeFlags & TransformFlags.ContainsDefaultValueAssignments) {
+                    transformFlags |= TransformFlags.AssertES6;
+                }
+
+                // An async function declaration is TypeScript syntax.
+                if (node.flags & NodeFlags.Async) {
+                    transformFlags |= TransformFlags.AssertTypeScript;
+                }
+
+                break;
+
+            case SyntaxKind.VariableDeclaration:
+                // A VariableDeclaration with a binding pattern is ES6 syntax.
+                if (isBindingPattern((<VariableDeclaration>node).name)) {
+                    transformFlags |= TransformFlags.AssertES6;
+                }
+
+                break;
+
+            case SyntaxKind.VariableDeclarationList:
+                // If a VariableDeclarationList is `let` or `const`, then it is ES6 syntax.
+                if (node.flags & NodeFlags.Let
+                    || node.flags & NodeFlags.Const) {
+                    transformFlags |= TransformFlags.AssertES6;
+                }
+
+                break;
+
+            case SyntaxKind.VariableStatement:
+                // If a VariableStatement is exported, then it is ES6 syntax.

```

Comment:
This should only be true at top-level (i.e. not inside `namespace` declarations)